File: backend/models/bpjs.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// BPJSConfig stores BPJS API configuration
type BPJSConfig struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	Key         string         `gorm:"uniqueIndex;not null;size:100" json:"key"`
	Value       string         `gorm:"type:text" json:"value"`
	Description string         `gorm:"type:text" json:"description"`
	IsEncrypted bool           `gorm:"default:false" json:"is_encrypted"`
	IsSecret    bool           `gorm:"default:false" json:"is_secret"` // If true, don't return value to frontend
}

// BPJSQueue stores BPJS online queue data
type BPJSQueue struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// BPJS Data
	KodeBooking    string    `gorm:"uniqueIndex;not null;size:50" json:"kode_booking"`
	NomorAntrean   string    `gorm:"not null;size:20" json:"nomor_antrean"`
	AngkaAntrean   int       `json:"angka_antrean"`
	NoKartu        string    `gorm:"not null;size:20;index" json:"no_kartu"`
	NIK            string    `gorm:"not null;size:20;index" json:"nik"`
	NamaPasien     string    `gorm:"not null;size:100" json:"nama_pasien"`
	NoHP           string    `gorm:"size:20" json:"no_hp"`
	TanggalPeriksa time.Time `gorm:"not null;index" json:"tanggal_periksa"`

	// Poli Info
	KodePoli   string `gorm:"not null;size:10;index" json:"kode_poli"`
	NamaPoli   string `gorm:"size:100" json:"nama_poli"`
	KodeDokter string `gorm:"size:20" json:"kode_dokter"`
	NamaDokter string `gorm:"size:100" json:"nama_dokter"`
	JamPraktek string `gorm:"size:20" json:"jam_praktek"`

	// Reference
	JenisKunjungan int    `gorm:"default:2" json:"jenis_kunjungan"` // 1=FKTP, 2=Internal, 3=Kontrol, 4=Antar RS
	NomorReferensi string `gorm:"size:50" json:"nomor_referensi"`
	PasienBaru     int    `gorm:"default:0" json:"pasien_baru"`

	// Status Tracking
	Status      string     `gorm:"default:'pending';size:20" json:"status"` // pending, checkedin, serving, completed, cancelled
	CheckedInAt *time.Time `json:"checked_in_at"`

	// Task Tracking (BPJS Task IDs)
	Task1At  *time.Time `json:"task_1_at"`  // Mulai tunggu admisi
	Task2At  *time.Time `json:"task_2_at"`  // Selesai tunggu admisi
	Task3At  *time.Time `json:"task_3_at"`  // Mulai tunggu poli
	Task4At  *time.Time `json:"task_4_at"`  // Dipanggil dokter
	Task5At  *time.Time `json:"task_5_at"`  // Mulai periksa
	Task6At  *time.Time `json:"task_6_at"`  // Selesai periksa
	Task7At  *time.Time `json:"task_7_at"`  // Mulai tunggu farmasi
	Task99At *time.Time `json:"task_99_at"` // Selesai/Batal

	// Link to SIMRS
	PatientID      *uint `json:"patient_id"`
	RegistrationID *uint `json:"registration_id"`
	VisitID        *uint `json:"visit_id"`
	RoomQueueID    *uint `json:"room_queue_id"`

	// Sync Status
	SyncStatus string     `gorm:"default:'synced';size:20" json:"sync_status"` // synced, pending, failed
	LastSyncAt *time.Time `json:"last_sync_at"`
	SyncError  string     `gorm:"type:text" json:"sync_error"`
	Notes      string     `gorm:"type:text" json:"notes"`

	// Relations
	Patient      *Patient      `gorm:"foreignKey:PatientID" json:"patient,omitempty"`
	Registration *Registration `gorm:"foreignKey:RegistrationID" json:"registration,omitempty"`
	Visit        *Visit        `gorm:"foreignKey:VisitID" json:"visit,omitempty"`
	RoomQueue    *RoomQueue    `gorm:"foreignKey:RoomQueueID" json:"room_queue,omitempty"`
}

// BPJSSyncLog stores API request/response logs
type BPJSSyncLog struct {
	ID        uint      `gorm:"primarykey" json:"id"`
	CreatedAt time.Time `json:"created_at"`

	// Request Info
	Endpoint    string `gorm:"not null;size:255;index" json:"endpoint"`
	Method      string `gorm:"not null;size:10" json:"method"`
	RequestBody string `gorm:"type:text" json:"request_body"`

	// Response Info
	ResponseCode int    `json:"response_code"`
	ResponseBody string `gorm:"type:text" json:"response_body"`

	// Status
	Status       string `gorm:"not null;size:20;index" json:"status"` // success, failed, timeout
	ErrorMessage string `gorm:"type:text" json:"error_message"`

	// Timing
	RequestAt  time.Time  `gorm:"not null;index" json:"request_at"`
	ResponseAt *time.Time `json:"response_at"`
	DurationMs int        `json:"duration_ms"`

	// Reference
	ReferenceType string `gorm:"size:50" json:"reference_type"` // bpjs_queue, sep, claim
	ReferenceID   *uint  `json:"reference_id"`
}

// BPJSPoliMapping maps SIMRS rooms to BPJS poli codes
type BPJSPoliMapping struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// SIMRS Room
	RoomID   uint   `gorm:"not null;uniqueIndex" json:"room_id"`
	RoomCode string `gorm:"size:20" json:"room_code"`
	RoomName string `gorm:"size:100" json:"room_name"`

	// BPJS Poli
	KodePoliBPJS string `gorm:"not null;size:10;index" json:"kode_poli_bpjs"`
	NamaPoliBPJS string `gorm:"size:100" json:"nama_poli_bpjs"`

	IsActive bool `gorm:"default:true" json:"is_active"`

	// Relations
	Room           *Room               `gorm:"foreignKey:RoomID" json:"room,omitempty"`
	DoctorMappings []BPJSDoctorMapping `gorm:"foreignKey:PoliMappingID" json:"doctor_mappings,omitempty"`
}

// BPJSDoctorMapping maps SIMRS doctor schedules to BPJS doctor codes
// Each doctor can have different BPJS codes for different poli
type BPJSDoctorMapping struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// Link to Poli Mapping
	PoliMappingID uint             `gorm:"not null;index" json:"poli_mapping_id"`
	PoliMapping   *BPJSPoliMapping `gorm:"foreignKey:PoliMappingID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"poli_mapping,omitempty"`

	// SIMRS Doctor Schedule
	DoctorScheduleID uint            `gorm:"index" json:"doctor_schedule_id"` // Optional - specific schedule
	DoctorSchedule   *DoctorSchedule `gorm:"foreignKey:DoctorScheduleID" json:"doctor_schedule,omitempty"`

	// SIMRS Employee (Doctor)
	EmployeeID   uint      `gorm:"not null;index" json:"employee_id"`
	Employee     *Employee `gorm:"foreignKey:EmployeeID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"employee,omitempty"`
	EmployeeName string    `gorm:"size:100" json:"employee_name"` // Cache for display

	// BPJS Doctor Info
	KodeDokterBPJS string `gorm:"not null;size:20;index" json:"kode_dokter_bpjs"`
	NamaDokterBPJS string `gorm:"size:100" json:"nama_dokter_bpjs"`

	// Schedule Info from BPJS (for display/validation)
	JadwalHari  string `gorm:"size:50" json:"jadwal_hari"` // e.g., "Senin, Rabu, Jumat"
	JamPraktek  string `gorm:"size:50" json:"jam_praktek"` // e.g., "08:00-12:00"
	KuotaJKN    int    `gorm:"default:0" json:"kuota_jkn"` // Kuota pasien JKN per hari
	KuotaNonJKN int    `gorm:"default:0" json:"kuota_non_jkn"`

	IsActive bool `gorm:"default:true" json:"is_active"`
}

// TableName returns the table name for BPJSConfig
func (BPJSConfig) TableName() string {
	return "bpjs_configs"
}

// TableName returns the table name for BPJSQueue
func (BPJSQueue) TableName() string {
	return "bpjs_queues"
}

// TableName returns the table name for BPJSSyncLog
func (BPJSSyncLog) TableName() string {
	return "bpjs_sync_logs"
}

// TableName returns the table name for BPJSPoliMapping
func (BPJSPoliMapping) TableName() string {
	return "bpjs_poli_mappings"
}

// TableName returns the table name for BPJSDoctorMapping
func (BPJSDoctorMapping) TableName() string {
	return "bpjs_doctor_mappings"
}

// BPJS Config Keys
const (
	BPJSConfigConsID          = "bpjs_cons_id"
	BPJSConfigSecretKey       = "bpjs_secret_key"
	BPJSConfigUserKey         = "bpjs_user_key"
	BPJSConfigKodePPK         = "bpjs_kode_ppk"
	BPJSConfigNamaPPK         = "bpjs_nama_ppk"
	BPJSConfigEnvironment     = "bpjs_environment"
	BPJSConfigBaseURLDev      = "bpjs_base_url_dev"
	BPJSConfigBaseURLProd     = "bpjs_base_url_prod"
	BPJSConfigSyncInterval    = "bpjs_sync_interval_minutes"
	BPJSConfigAutoSyncEnabled = "bpjs_auto_sync_enabled"
)

// Default BPJS Config values
var DefaultBPJSConfigs = []BPJSConfig{
	{Key: BPJSConfigConsID, Value: "", Description: "Consumer ID dari BPJS", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigSecretKey, Value: "", Description: "Secret Key dari BPJS", IsEncrypted: true, IsSecret: true},
	{Key: BPJSConfigUserKey, Value: "", Description: "User Key dari BPJS", IsEncrypted: true, IsSecret: true},
	{Key: BPJSConfigKodePPK, Value: "", Description: "Kode Faskes/PPK", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigNamaPPK, Value: "", Description: "Nama Faskes", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigEnvironment, Value: "development", Description: "Environment: development atau production", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigBaseURLDev, Value: "https://apijkn-dev.bpjs-kesehatan.go.id", Description: "Base URL Development", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigBaseURLProd, Value: "https://apijkn.bpjs-kesehatan.go.id", Description: "Base URL Production", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigSyncInterval, Value: "5", Description: "Interval sinkronisasi dalam menit", IsEncrypted: false, IsSecret: false},
	{Key: BPJSConfigAutoSyncEnabled, Value: "false", Description: "Enable auto sync", IsEncrypted: false, IsSecret: false},
}

File: frontend/src/lib/api/bpjs.ts
import { api } from './client';

// BPJS Config Types
export interface BPJSConfigValue {
  value: string;
  has_value: boolean;
  description: string;
  is_secret: boolean;
  is_encrypted: boolean;
}

export interface BPJSConfigMap {
  bpjs_cons_id?: BPJSConfigValue;
  bpjs_secret_key?: BPJSConfigValue;
  bpjs_user_key?: BPJSConfigValue;
  bpjs_kode_ppk?: BPJSConfigValue;
  bpjs_nama_ppk?: BPJSConfigValue;
  bpjs_environment?: BPJSConfigValue;
  bpjs_base_url_dev?: BPJSConfigValue;
  bpjs_base_url_prod?: BPJSConfigValue;
  bpjs_sync_interval_minutes?: BPJSConfigValue;
  bpjs_auto_sync_enabled?: BPJSConfigValue;
  [key: string]: BPJSConfigValue | undefined;
}

export interface BPJSConnectionTestResult {
  success: boolean;
  message?: string;
  error?: string;
  environment?: string;
  base_url?: string;
  response_code?: number;
  response_time?: string;
  response_body?: string;
  duration?: number;
  poli_count?: number;
}

export interface BPJSFingerprintLaunchRequest {
  executable_path?: string;
  username?: string;
  password?: string;
  auto_submit?: boolean;
  bpjs_card_number?: string;
  patient_nik?: string;
}

export interface BPJSFingerprintLaunchResponse {
  message: string;
  data: {
    executable_path: string;
    auto_submit: boolean;
    bpjs_card_number?: string;
    patient_nik?: string;
  };
}

export interface BPJSSyncLog {
  id: number;
  endpoint: string;
  method: string;
  request_body: string;
  response_code: number;
  response_body: string;
  status: 'success' | 'failed' | 'timeout';
  error_message: string;
  request_at: string;
  response_at: string;
  duration_ms: number;
  reference_type: string;
  reference_id: number;
  created_at: string;
}

export interface BPJSSyncStats {
  total_requests: number;
  success_requests: number;
  failed_requests: number;
  avg_duration_ms: number;
}

export interface BPJSApotekDPHOItem {
  kodeobat: string;
  namaobat: string;
  prb?: string | boolean | number | null;
  kronis?: string | boolean | number | null;
  kemo?: string | boolean | number | null;
  harga?: string | null;
  restriksi?: string | null;
  generik?: string | null;
  aktif?: string | null;
  sedia?: string | null;
  stok?: string | null;
  [key: string]: unknown;
}

export interface BPJSApotekDPHOData {
  list?: BPJSApotekDPHOItem[];
}

// Referensi BPJS Types
export interface BPJSReferensiPoli {
  kdpoli: string;
  nmpoli: string;
}

export interface BPJSReferensiDokter {
  kodedokter: number;
  namadokter: string;
}

export interface BPJSJadwalDokter {
  kodedokter: number;
  namadokter: string;
  kodepoli: string;
  namapoli: string;
  hari: number;
  namahari: string;
  jadwal: string;
  libur: number;
  kapasitaspasien: number;
}

// Mapping Types
export interface BPJSPoliMapping {
  id: number;
  room_id: number;
  room_code: string;
  room_name: string;
  kode_poli_bpjs: string;
  nama_poli_bpjs: string;
  is_active: boolean;
  room?: {
    id: number;
    code: string;
    name: string;
  };
  doctor_mappings?: BPJSDoctorMapping[];
  created_at: string;
  updated_at: string;
}

export interface BPJSDoctorMapping {
  id: number;
  poli_mapping_id: number;
  poli_mapping?: BPJSPoliMapping;
  doctor_schedule_id?: number;
  employee_id: number;
  employee_name: string;
  employee?: {
    id: number;
    nama_lengkap: string;
    nip?: string;
  };
  kode_dokter_bpjs: string;
  nama_dokter_bpjs: string;
  jadwal_hari: string;
  jam_praktek: string;
  kuota_jkn: number;
  kuota_non_jkn: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePoliMappingRequest {
  room_id: number;
  kode_poli_bpjs: string;
  nama_poli_bpjs: string;
}

export interface UpdatePoliMappingRequest {
  kode_poli_bpjs?: string;
  nama_poli_bpjs?: string;
  is_active?: boolean;
}

export interface CreateDoctorMappingRequest {
  poli_mapping_id: number;
  employee_id: number;
  doctor_schedule_id?: number;
  kode_dokter_bpjs: string;
  nama_dokter_bpjs: string;
  jadwal_hari?: string;
  jam_praktek?: string;
  kuota_jkn?: number;
  kuota_non_jkn?: number;
}

export interface UpdateDoctorMappingRequest {
  kode_dokter_bpjs?: string;
  nama_dokter_bpjs?: string;
  jadwal_hari?: string;
  jam_praktek?: string;
  kuota_jkn?: number;
  kuota_non_jkn?: number;
  is_active?: boolean;
}

export const bpjsApi = {
  // Config
  getConfig: () => 
    api.get<{ data: BPJSConfigMap }>('/bpjs/config'),
  
  updateConfig: (data: Record<string, string>) => 
    api.put<{ message: string }>('/bpjs/config', data),
  
  initConfig: () => 
    api.post<{ message: string }>('/bpjs/config/init'),
  
  testConnection: () => 
    api.get<BPJSConnectionTestResult>('/bpjs/test-connection'),

  launchFingerprintApp: (data: BPJSFingerprintLaunchRequest) =>
    api.post<BPJSFingerprintLaunchResponse>('/bpjs/fingerprint/launch', data),

  // Logs
  getLogs: (params?: { status?: string; date?: string; endpoint?: string; limit?: number }) => 
    api.get<{ data: BPJSSyncLog[] }>('/bpjs/logs', { params }),
  
  getStats: () => 
    api.get<{ data: BPJSSyncStats }>('/bpjs/logs/stats'),

  // Referensi BPJS
  getReferensiPoli: () =>
    api.get<{ data: BPJSReferensiPoli[]; count: number }>('/bpjs/referensi/poli'),
  
  getReferensiDokter: () =>
    api.get<{ data: BPJSReferensiDokter[]; count: number }>('/bpjs/referensi/dokter'),
  
  getJadwalDokter: (kodePoli: string, tanggal?: string) =>
    api.get<{ data: BPJSJadwalDokter[]; count: number }>('/bpjs/referensi/jadwal-dokter', {
      params: { kode_poli: kodePoli, tanggal }
    }),

  // Poli Mapping
  getPoliMappings: (params?: { room_id?: number; is_active?: boolean }) => 
    api.get<{ data: BPJSPoliMapping[] }>('/bpjs/mapping/poli', { params }),
  
  createPoliMapping: (data: CreatePoliMappingRequest) => 
    api.post<{ data: BPJSPoliMapping }>('/bpjs/mapping/poli', data),
  
  updatePoliMapping: (id: number, data: UpdatePoliMappingRequest) => 
    api.put<{ data: BPJSPoliMapping }>(`/bpjs/mapping/poli/${id}`, data),
  
  deletePoliMapping: (id: number) => 
    api.delete<{ message: string }>(`/bpjs/mapping/poli/${id}`),

  // Doctor Mapping
  getDoctorMappings: (params?: { poli_mapping_id?: number; employee_id?: number; is_active?: boolean }) =>
    api.get<{ data: BPJSDoctorMapping[] }>('/bpjs/mapping/dokter', { params }),
  
  createDoctorMapping: (data: CreateDoctorMappingRequest) =>
    api.post<{ data: BPJSDoctorMapping }>('/bpjs/mapping/dokter', data),
  
  updateDoctorMapping: (id: number, data: UpdateDoctorMappingRequest) =>
    api.put<{ data: BPJSDoctorMapping }>(`/bpjs/mapping/dokter/${id}`, data),
  
  deleteDoctorMapping: (id: number) =>
    api.delete<{ message: string }>(`/bpjs/mapping/dokter/${id}`),
  
  syncDoctorFromReferensi: () =>
    api.post<{ message: string; data: Record<string, unknown> }>('/bpjs/mapping/dokter/sync'),

  // BPJS Queue (Antrian MJKN)
  getQueues: (params?: { date?: string; status?: string; kode_poli?: string; limit?: number }) =>
    api.get<{ data: BPJSQueue[] }>('/bpjs/queue', { params }),
  
  getQueueByRegistration: (registrationId: number) =>
    api.get<{ data: BPJSQueue }>(`/bpjs/queue/registration/${registrationId}`),
  
  getQueueByVisit: (visitId: number) =>
    api.get<{ data: BPJSQueue }>(`/bpjs/queue/visit/${visitId}`),
  
  activateQueueCheckin: (queueId: number) =>
    api.post<{ message: string; data: BPJSQueue }>(`/bpjs/queue/${queueId}/checkin`),
  
  // Manual send task to BPJS (for monitoring page)
  // waktu: timestamp in milliseconds
  sendTaskManual: (queueId: number, taskId: number, waktu: number) =>
    api.post<{ 
      success: boolean; 
      message: string; 
      response_code: number; 
      response_msg: string;
      waktu_sent: number;
    }>(`/bpjs/queue/${queueId}/send-task`, { task_id: taskId, waktu }),
  
  // Retry AddAntrean to BPJS
  retryAddAntrean: (queueId: number) =>
    api.post<{
      success: boolean;
      response_code: number;
      response_msg: string;
      data: BPJSQueue;
    }>(`/bpjs/queue/${queueId}/retry-add`),

  // Cancel BPJS queue (from internal system)
  cancelQueue: (queueId: number) =>
    api.post<{ message: string; data: BPJSQueue }>(`/bpjs/queue/${queueId}/cancel`),

  // === Antrian Online (direct BPJS API calls) ===

  // Get pendaftaran antrean by date from BPJS Antrian Online
  getPendaftaranAntrean: (tanggal: string) =>
    api.get<{ data: BPJSPendaftaranAntreanItem[] }>(`/bpjs/antrean/pendaftaran/${tanggal}`),

  // Batal antrean at BPJS Antrian Online
  batalAntrean: (kodebooking: string, keterangan: string) =>
    api.post<{ message: string; response_code: number; response_msg: string }>(
      '/bpjs/antrean/batal', { kodebooking, keterangan }
    ),

  // Get pendaftaran detail by kode booking
  getPendaftaranByKodeBooking: (kodebooking: string) =>
    api.get<{ data: BPJSPendaftaranAntreanItem[] }>(`/bpjs/antrean/pendaftaran-detail/${kodebooking}`),

  // Get list task by kode booking
  getListTask: (kodebooking: string) =>
    api.post<{ data: BPJSListTaskItem[] }>('/bpjs/antrean/getlisttask', { kodebooking }),

  // === I-Care ===
  
  // Validate I-Care: returns URL to open BPJS I-Care web interface
  icareValidate: (visitId: number) =>
    api.post<{ url: string; message: string }>(`/bpjs/icare/validate/${visitId}`),
  
  // Validate I-Care manual: input nomor kartu dan kode dokter langsung
  icareValidateManual: (noKartu: string, kodeDokter: number) =>
    api.post<{ url: string; message: string }>('/bpjs/icare/validate-manual', { no_kartu: noKartu, kode_dokter: kodeDokter }),

  // === APLICARE (Ketersediaan Tempat Tidur) ===

  // Referensi kelas kamar dari BPJS
  aplicareGetRefKelas: () =>
    api.get<{ data: AplicareRefKelasItem[] }>('/bpjs/aplicare/ref-kelas'),

  // Baca ketersediaan tempat tidur dari BPJS
  aplicareReadBed: (start = 1, limit = 100) =>
    api.get<{ data: AplicareBedItem[] }>(`/bpjs/aplicare/bed?start=${start}&limit=${limit}`),

  // Daftar ruangan rawat inap SIMRS (yang punya bed)
  aplicareGetRooms: () =>
    api.get<{ data: AplicareRoom[] }>('/bpjs/aplicare/rooms'),

  // Daftarkan ruangan baru ke Aplicare
  aplicareCreateRoom: (payload: AplicareCreateRoomPayload | number) =>
    api.post<{ message: string; data: AplicareBedRequest }>('/bpjs/aplicare/bed/create', typeof payload === 'number' ? { room_id: payload } : payload),

  // Update ketersediaan tempat tidur
  aplicareUpdateRoom: (roomId: number) =>
    api.post<{ message: string; data: AplicareBedRequest }>('/bpjs/aplicare/bed/update', { room_id: roomId }),

  // Hapus seluruh ruangan beserta unitnya dari Aplicare
  aplicareDeleteRoom: (roomId: number) =>
    api.post<{ message: string }>('/bpjs/aplicare/bed/delete', { room_id: roomId }),

  // === APOTEK ONLINE ===

  apotekGetReferensiDPHO: () =>
    api.get<{ data: BPJSApotekDPHOData | BPJSApotekDPHOItem[]; cached?: boolean; warning?: string }>('/bpjs/apotek/referensi/dpho'),

  apotekGetReferensiPoli: (parameter: string) =>
    api.get<{ data: unknown }>(`/bpjs/apotek/referensi/poli/${encodeURIComponent(parameter)}`),

  apotekGetFasilitasKesehatan: (jenisFaskes: string, namaFaskes: string) =>
    api.get<{ data: unknown }>(
      `/bpjs/apotek/referensi/ppk/${encodeURIComponent(jenisFaskes)}/${encodeURIComponent(namaFaskes)}`,
    ),

  apotekGetSettingApotek: (kodeApotek: string) =>
    api.get<{ data: unknown }>(`/bpjs/apotek/referensi/settingppk/${encodeURIComponent(kodeApotek)}`),

  apotekGetSpesialistik: () =>
    api.get<{ data: unknown }>('/bpjs/apotek/referensi/spesialistik'),

  apotekGetReferensiObat: (kodeJenisObat: string, tglResep: string, filter: string) =>
    api.get<{ data: unknown }>('/bpjs/apotek/referensi/obat', {
      params: {
        kodeJenisObat,
        tglResep,
        filter,
      },
    }),

  apotekInsertObatNonRacikan: (payload: Record<string, unknown>) =>
    api.post<{ data: unknown }>('/bpjs/apotek/obat/non-racikan', payload),

  apotekInsertObatRacikan: (payload: Record<string, unknown>) =>
    api.post<{ data: unknown }>('/bpjs/apotek/obat/racikan', payload),

  apotekUpdateStokObat: (payload: Record<string, unknown>) =>
    api.post<{ data: unknown }>('/bpjs/apotek/obat/stok', payload),

  apotekHapusPelayananObat: (payload: Record<string, unknown>) =>
    api.delete<{ data: unknown }>('/bpjs/apotek/pelayanan/obat', { data: payload }),

  apotekGetDaftarPelayananObat: (noKunjungan: string) =>
    api.get<{ data: unknown }>(`/bpjs/apotek/obat/daftar/${encodeURIComponent(noKunjungan)}`),

  apotekGetRiwayatPelayananObat: (tglAwal: string, tglAkhir: string, noKartu: string) =>
    api.get<{ data: unknown }>(
      `/bpjs/apotek/riwayat/${encodeURIComponent(tglAwal)}/${encodeURIComponent(tglAkhir)}/${encodeURIComponent(noKartu)}`,
    ),

  apotekSimpanResep: (payload: Record<string, unknown>) =>
    api.post<{ data: unknown }>('/bpjs/apotek/resep/simpan', payload),

  apotekHapusResep: (payload: Record<string, unknown>) =>
    api.delete<{ data: unknown }>('/bpjs/apotek/resep', { data: payload }),

  apotekDaftarResep: (payload: Record<string, unknown>) =>
    api.post<{ data: unknown }>('/bpjs/apotek/resep/daftar', payload),

  apotekCariKunjunganBySEP: (noSEP: string) =>
    api.get<{ data: unknown }>(`/bpjs/apotek/sep/${encodeURIComponent(noSEP)}`),

  apotekGetDataKlaim: (bulan: string, tahun: string, jenisObat: string, status: string) =>
    api.get<{ data: unknown }>(
      `/bpjs/apotek/monitoring/klaim/${encodeURIComponent(bulan)}/${encodeURIComponent(tahun)}/${encodeURIComponent(jenisObat)}/${encodeURIComponent(status)}`,
    ),

  apotekGetRekapPesertaPRB: (tahun: string, bulan: string) =>
    api.get<{ data: unknown }>(`/bpjs/apotek/prb/rekap/${encodeURIComponent(tahun)}/${encodeURIComponent(bulan)}`),
};

// BPJS Queue Types (Antrian MJKN)
export interface BPJSQueue {
  id: number;
  kode_booking: string;
  nomor_antrean: string;
  angka_antrean: number;
  tanggal_periksa: string;
  jam_praktek: string;
  kode_poli: string;
  nama_poli: string;
  kode_dokter: string;
  nama_dokter: string;
  jenis_pasien: string;
  no_kartu: string;
  nik: string;
  no_hp: string;
  no_rm: string;
  nama_pasien: string;
  jenis_kunjungan: number;
  nomor_referensi: string;
  estimasi_dilayani: number;
  status: string;
  keterangan: string;
  waktu_checkin?: string;
  waktu_batal?: string;
  
  // Farmasi
  nomor_antrean_farmasi: number;
  status_farmasi: string;
  
  // Farmasi BPJS Task Buffer
  // Jika ada: farmasi sudah di-order/selesai tapi Task 5/6 belum terkirim, menunggu auto-chain
  farmasi_ready_at?: string;
  farmasi_jenis_resep?: string;
  farmasi_selesai_at?: string;
  
  // Task tracking
  task1_at?: string;
  task2_at?: string;
  task3_at?: string;
  task4_at?: string;
  task5_at?: string;
  task6_at?: string;
  task7_at?: string;
  
  // SIMRS Relations
  patient_id?: number;
  registration_id?: number;
  visit_id?: number;
  room_queue_id?: number;
  room_id?: number;
  poli_mapping_id?: number;
  doctor_mapping_id?: number;
  
  // Sync
  sync_status: string;
  sync_error?: string;
  last_sync_at?: string;
  
  // AddAntrean tracking
  add_antrean_sent: boolean;
  add_antrean_code: number;
  add_antrean_msg?: string;
  
  // Relations
  patient?: {
    id: number;
    no_rm: string;
    nama_lengkap: string;
    nik: string;
  };
  registration?: {
    id: number;
    registration_number: string;
    status: string;
  };
  visit?: {
    id: number;
    visit_number: string;
    status: string;
  };
  room_queue?: {
    id: number;
    queue_number: string;
    status: string;
  };
  room?: {
    id: number;
    name: string;
    code: string;
  };
  
  created_at: string;
  updated_at: string;
}

// BPJS Pendaftaran Antrean Item (from Antrian Online API)
export interface BPJSPendaftaranAntreanItem {
  kodebooking: string;
  tanggal: string;
  kodepoli: string;
  kodedokter: number;
  jampraktek: string;
  nik: string;
  nokapst: string;
  nohp: string;
  norekammedis: string;
  jeniskunjungan: number;
  nomorreferensi: string;
  sumberdata: string;
  ispeserta: boolean | number;
  noantrean: string;
  estimasidilayani: number;
  createdtime: number;
  status: string;
}

// BPJS List Task Item (from Antrian Online getlisttask API)
export interface BPJSListTaskItem {
  taskid: number;
  taskname: string;
  waktu: string;
  wakturs: string;
  wpidr: string;
}

// === APLICARE Types ===

export interface AplicareRefKelasItem {
  kodekelas: string;
  namakelas: string;
}

export interface AplicareBedItem {
  kodekelas: string;
  namakelas: string;
  koderuang: string;
  namaruang: string;
  kapasitas: number;
  tersedia: number;
  tersediapria: number;
  tersediawanita: number;
  tersediapriawanita: number;
}

export interface AplicareBedRequest {
  kodekelas: string;
  koderuang: string;
  namaruang: string;
  kapasitas: string;
  tersedia: string;
  tersediapria: string;
  tersediawanita: string;
  tersediapriawanita: string;
}

export interface AplicareCreateRoomPayload extends Partial<AplicareBedRequest> {
  room_id: number;
  sync_mode?: boolean; // true = update existing too; false = only create new
}

export interface AplicareRoom {
  id: number;
  code: string;
  name: string;
  room_class: string;
  room_type: string;
  has_bed: boolean;
  is_active: boolean;
  total_beds: number;
  available_beds: number;
}
